url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/core/SimpleThreadPool.java

```diff
@@ -0,0 +1,567 @@
+        busyWorkers.add(wt);
+        workers.add(wt);
+        wt.start();
+      }
+      nextRunnableLock.notifyAll();
+      handoffPending = false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public int blockForAvailableThreads() {
+
+    synchronized (nextRunnableLock) {
+
+      while ((availWorkers.size() < 1 || handoffPending) && !isShutdown) {
+        try {
+          nextRunnableLock.wait(500);
+        } catch (InterruptedException ignore) {
+        }
+      }
+
+      return availWorkers.size();
+    }
+  }
+
+  private void makeAvailable(WorkerThread wt) {
+
+    synchronized (nextRunnableLock) {
+      if (!isShutdown) {
+        availWorkers.add(wt);
+      }
+      busyWorkers.remove(wt);
+      nextRunnableLock.notifyAll();
+    }
+  }
+
+  private void clearFromBusyWorkersList(WorkerThread wt) {
+
+    synchronized (nextRunnableLock) {
+      busyWorkers.remove(wt);
+      nextRunnableLock.notifyAll();
+    }
+  }
+
+  /*
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ WorkerThread Class.
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+
+  /**
+   * <p>
+   * A Worker loops, waiting to execute tasks.
+   * </p>
+   */
+  private class WorkerThread extends Thread {
+
+    // A flag that signals the WorkerThread to terminate.
+    private boolean run = true;
+
+    private SimpleThreadPool tp;
+
+    private Runnable runnable = null;
+
+    private boolean runOnce = false;
+
+    /**
+     * <p>
+     * Create a worker thread and start it. Waiting for the next Runnable, executing it, and waiting for the next Runnable, until the shutdown flag is
+     * set.
+     * </p>
+     */
+    private WorkerThread(SimpleThreadPool tp, ThreadGroup threadGroup, String name, int prio, boolean isDaemon) {
+
+      this(tp, threadGroup, name, prio, isDaemon, null);
+    }
+
+    /**
+     * <p>
+     * Create a worker thread, start it, execute the runnable and terminate the thread (one time execution).
+     * </p>
+     */
+    private WorkerThread(SimpleThreadPool tp, ThreadGroup threadGroup, String name, int prio, boolean isDaemon, Runnable runnable) {
+
+      super(threadGroup, name);
+      this.tp = tp;
+      this.runnable = runnable;
+      if (runnable != null) {
+        runOnce = true;
+      }
+      setPriority(prio);
+      setDaemon(isDaemon);
+    }
+
+    /**
+     * <p>
+     * Signal the thread that it should terminate.
+     * </p>
+     */
+    void shutdown() {
+
+      synchronized (this) {
+        run = false;
+      }
+    }
+
+    public void run(Runnable newRunnable) {
+
+      synchronized (this) {
+        if (runnable != null) {
+          throw new IllegalStateException("Already running a Runnable!");
+        }
+
+        runnable = newRunnable;
+        this.notifyAll();
+      }
+    }
+
+    /**
+     * <p>
+     * Loop, executing targets as they are received.
+     * </p>
+     */
+    @Override
+    public void run() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.